File: pantomime-lib/Framework/Pantomime/TCPConnection.m
//
//  TCPConnection.m
//  Pantomime
//
//  Created by Dirk Zimmermann on 12/04/16.
//  Copyright © 2016 pEp Security S.A. All rights reserved.
//

#import "TCPConnection.h"

static NSString *comp = @"TCPConnection";

@interface TCPConnection ()

@property (nonatomic) BOOL connected;
@property (nonatomic, strong) NSString *name;
@property (nonatomic) uint32_t port;
@property (nonatomic) ConnectionTransport transport;
@property (nonatomic, strong) NSInputStream *readStream;
@property (nonatomic, strong) NSOutputStream *writeStream;
@property (nonatomic, strong) NSMutableSet<NSStream *> *openConnections;

@end

@implementation TCPConnection

- (instancetype)initWithName:(NSString *)theName port:(unsigned int)thePort
         transport:(ConnectionTransport)transport background:(BOOL)theBOOL
{
    if (self = [super init]) {
        _openConnections = [[NSMutableSet alloc] init];
        _connected = NO;
        _name = [theName copy];
        _port = thePort;
        _transport = transport;
        NSAssert(theBOOL, @"TCPConnection only supports background mode");
    }
    return self;
}

- (void)startTLS
{
    [self.readStream setProperty:NSStreamSocketSecurityLevelNegotiatedSSL
                          forKey:NSStreamSocketSecurityLevelKey];
    [self.writeStream setProperty:NSStreamSocketSecurityLevelNegotiatedSSL
                           forKey:NSStreamSocketSecurityLevelKey];
}

- (void)setupStream:(NSStream *)stream
{
    stream.delegate = self;
    switch (self.transport) {
        case ConnectionTransportPlain:
        case ConnectionTransportStartTLS:
            [stream setProperty:NSStreamSocketSecurityLevelNone
                         forKey:NSStreamSocketSecurityLevelKey];
            break;
        case ConnectionTransportTLS:
            [stream setProperty:NSStreamSocketSecurityLevelNegotiatedSSL
                         forKey:NSStreamSocketSecurityLevelKey];
            break;
    }
    [stream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
    [stream open];
}

- (void)closeAndRemoveStream:(NSStream *)stream
{
    if (stream) {
        [stream close];
        [self.openConnections removeObject:stream];
        if (stream == self.readStream) {
            self.readStream = nil;
        } else if (stream == self.writeStream) {
            self.writeStream = nil;
        }
    }
}

#pragma mark -- CWConnection

- (BOOL)isConnected
{
    return _connected;
}

- (void)close
{
    [self closeAndRemoveStream:self.readStream];
    [self closeAndRemoveStream:self.writeStream];
    self.connected = NO;
    [self.delegate receivedEvent:nil type:ET_EDESC extra:nil forMode:nil];
}

- (NSString *)bufferToString:(unsigned char *)buf length:(NSInteger)length
{
    if (length) {
        NSData *data = [NSData dataWithBytes:buf length:length];
        NSString *string = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        return string;
    } else {
        return @"";
    }
}

- (NSInteger)read:(unsigned char *)buf length:(NSInteger)len
{
    if (![self.readStream hasBytesAvailable]) {
        return -1;
    }
    NSInteger count = [self.readStream read:buf maxLength:len];
    NSLog(@"read %d: \"%@\"", count, [self bufferToString:buf length:len]);
    return count;
}

- (NSInteger) write:(unsigned char *)buf length:(NSInteger)len
{
    if (![self.writeStream hasSpaceAvailable]) {
        return -1;
    }
    NSInteger count = [self.writeStream write:buf maxLength:len];
    NSLog(@"wrote %d: \"%@\"", count, [self bufferToString:buf length:len]);
    return count;
}

- (void)connect
{
    CFReadStreamRef readStream = nil;
    CFWriteStreamRef writeStream = nil;
    CFStreamCreatePairWithSocketToHost(kCFAllocatorDefault, (__bridge CFStringRef) self.name,
                                       self.port, &readStream, &writeStream);

    NSAssert(readStream != nil, @"Could not create input stream");
    NSAssert(writeStream != nil, @"Could not create output stream");

    if (readStream != nil && writeStream != nil) {
        self.readStream = CFBridgingRelease(readStream);
        self.writeStream = CFBridgingRelease(writeStream);
        [self setupStream:self.readStream];
        [self setupStream:self.writeStream];
    }
}

- (BOOL)canWrite
{
    return [self.writeStream hasSpaceAvailable];
}

#pragma mark -- NSStreamDelegate

- (void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode
{
    switch (eventCode) {
        case NSStreamEventNone:
            NSLog(@"NSStreamEventNone");
            break;
        case NSStreamEventOpenCompleted:
            NSLog(@"NSStreamEventOpenCompleted");
            [self.openConnections addObject:aStream];
            if (self.openConnections.count == 2) {
                NSLog(@"connectionEstablished");
                self.connected = YES;
                [self.delegate connectionEstablished];
            }
            break;
        case NSStreamEventHasBytesAvailable:
            NSLog(@"NSStreamEventHasBytesAvailable");
            [self.delegate receivedEvent:nil type:ET_RDESC extra:nil forMode:nil];
            break;
        case NSStreamEventHasSpaceAvailable:
            NSLog(@"NSStreamEventHasSpaceAvailable");
            [self.delegate receivedEvent:nil type:ET_WDESC extra:nil forMode:nil];
            break;
        case NSStreamEventErrorOccurred:
            NSLog(@"NSStreamEventErrorOccurred");
            [self.delegate receivedEvent:nil type:ET_RDESC extra:nil forMode:nil];
            break;
        case NSStreamEventEndEncountered:
            NSLog(@"NSStreamEventEndEncountered");
            [self.delegate receivedEvent:nil type:ET_RDESC extra:nil forMode:nil];
            break;
    }
}

@end

File: pantomime-lib/Framework/Pantomime/Extensions/TCPConnection.swift
//
//  TCPConnection.swift
//  PantomimeMailOSX
//
//  Created by Dirk Zimmermann on 06/04/16.
//  Copyright © 2016 pEp Security S.A. All rights reserved.
//

import Foundation

@objc public class TCPConnection: NSObject {
    let comp = "TCPConnection"

    var connected = false
    let name: String
    let port: UInt32
    let transport: ConnectionTransport
    var readStream: NSInputStream? = nil
    var writeStream: NSOutputStream? = nil
    var openConnections = Set<NSStream>()

    /** Required from CWConnection */
    public var delegate: CWConnectionDelegate?

    /** Required from CWConnection */
    public required init(name: String, port: UInt32,
                         transport: ConnectionTransport,
                         background theBOOL: Bool) {
        assert(theBOOL == true, "Only asynchronous connections in background are supported")
        self.name = name
        self.port = port
        self.transport = transport
        super.init()
    }

    public func startTLS() {
        readStream!.setProperty(NSStreamSocketSecurityLevelNone,
                                forKey: NSStreamSocketSecurityLevelKey)
        writeStream!.setProperty(NSStreamSocketSecurityLevelNone,
                                 forKey: NSStreamSocketSecurityLevelKey)
        readStream?.open()
        writeStream?.open()
    }

    func setupStream(stream: NSStream?) {
        stream?.delegate = self
        switch transport {
        case .Plain:
            stream?.setProperty(NSStreamSocketSecurityLevelNone,
                                forKey: NSStreamSocketSecurityLevelKey)
        case .StartTLS:
            stream?.setProperty(NSStreamSocketSecurityLevelNone,
                                forKey: NSStreamSocketSecurityLevelKey)
        case .TLS:
            stream?.setProperty(NSStreamSocketSecurityLevelNegotiatedSSL,
                                forKey: NSStreamSocketSecurityLevelKey)
        }
        stream?.scheduleInRunLoop(NSRunLoop.currentRunLoop(), forMode: NSDefaultRunLoopMode)
        stream?.open()
    }

    func closeAndRemoveStream(stream: NSStream?) {
        if let theStream = stream {
            theStream.close()
            openConnections.remove(theStream)
            if theStream == readStream {
                readStream = nil
            } else if theStream == writeStream {
                writeStream = nil
            }
        }
    }

    func writeOutBuffer() -> Int {
        return 0
    }
}

extension TCPConnection: CWConnection {
    public func isConnected() -> Bool {
        return connected
    }

    public func close() {
        closeAndRemoveStream(readStream)
        closeAndRemoveStream(writeStream)
        connected = false
        delegate?.receivedEvent(nil, type: ET_EDESC, extra: nil, forMode: nil)
    }

    func uint8BytesToString(bytes: UnsafeMutablePointer<UInt8>, length: Int) -> String {
        if length >= 0 {
            let data = NSData(bytes: bytes, length: Int(length))
            let string = NSString(data: data, encoding: NSUTF8StringEncoding)
            return string as! String
        } else {
            return ""
        }
    }

    public func read(buf: UnsafeMutablePointer<UInt8>, length: Int) -> Int {
        if readStream?.hasBytesAvailable == false {
            return -1
        }
        let count = readStream!.read(buf, maxLength: length)
        Log.info(comp, content: "read \(count): \"\(uint8BytesToString(buf, length: count))\"")
        return count
    }

    public func write(buf: UnsafeMutablePointer<UInt8>, length: Int) -> Int {
        if writeStream?.hasSpaceAvailable == false {
            return -1
        }
        let count = writeStream!.write(buf, maxLength: length)
        Log.info(comp, content: "wrote \"\(count): \(uint8BytesToString(buf, length: count))\"")
        return count
    }

    public func connect() {
        var readStream:  Unmanaged<CFReadStream>? = nil
        var writeStream: Unmanaged<CFWriteStream>? = nil
        CFStreamCreatePairWithSocketToHost(kCFAllocatorDefault, name, port, &readStream,
                                           &writeStream)

        assert(readStream != nil, "Could not create reading stream")
        assert(writeStream != nil, "Could not create writing stream")

        if readStream != nil || writeStream != nil {
            self.readStream = readStream?.takeRetainedValue()
            self.writeStream = writeStream?.takeRetainedValue()
            self.setupStream(self.readStream)
            self.setupStream(self.writeStream)
        }
    }

    public func canWrite() -> Bool {
        return writeStream!.hasSpaceAvailable
    }

}

extension TCPConnection: NSStreamDelegate {
    public func stream(aStream: NSStream,
                       handleEvent eventCode: NSStreamEvent) {
        switch eventCode {
        case NSStreamEvent.None:
            Log.info(comp, content: "\(aStream) None")
        case NSStreamEvent.OpenCompleted:
            openConnections.insert(aStream)
            Log.info(comp, content: "\(aStream) OpenCompleted")
            if openConnections.count == 2 {
                connected = true
                Log.info(comp, content: "connectionEstablished")
                delegate?.connectionEstablished()
            }
        case NSStreamEvent.HasBytesAvailable:
            Log.info(comp, content: "HasBytesAvailable")
            delegate?.receivedEvent(nil, type: ET_RDESC, extra: nil, forMode: nil)
        case NSStreamEvent.HasSpaceAvailable:
            Log.info(comp, content: "HasSpaceAvailable")
            delegate?.receivedEvent(nil, type: ET_WDESC, extra: nil, forMode: nil)
        case NSStreamEvent.ErrorOccurred:
            Log.info(comp, content: "ErrorOccurred")
            delegate?.receivedEvent(nil, type: ET_RDESC, extra: nil, forMode: nil)
        case NSStreamEvent.EndEncountered:
            Log.info(comp, content: "EndEncountered")
            delegate?.receivedEvent(nil, type: ET_RDESC, extra: nil, forMode: nil)
        default:
            Log.info(comp, content: "eventCode \(eventCode)")
            delegate?.receivedEvent(nil, type: ET_EDESC, extra: nil, forMode: nil)
        }
    }

}
